### ISSUE TITLE
UTF-8 converted to Latin1 for output with format incorrect output

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9295]
Migrated from [rt.perl.org#53054](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=53054) (status was 'resolved')

Searchable as RT53054$

**Comment by p5pRT** [comment]
### From mdkoehn@yahoo.com

#### Created by mdkoehn@yahoo.com



I'm reading Unicode data from a file\, and have set the output filehandle to be ISO\-8859\-1 \(Latin1\)\.  
I'm using a format to format the output\.  Evertything is fine in the beginning\.  When I hit the first
line with an accented character\, the two spaces that I use as column dividers change into "\<\<" for the
rest of the program run\.  Here's an example&#8203;:

1234567890123    Grisham\, John     The chamber \[audiobook\] / John Grisham
1234567890123                      Flic story \[DVD\] / Pathé Renn production ; St
1234567890123    Tempelman\-Kluit\,\<\<Yukon / by Anne Tempelman\-Kluit ; consultants
1234567890123    Wood\, Tim       \<\<Motor racing / Tim Wood ; photographs\, Chris

The first line is fine\.  The second line has an accented character \(é\)\.  Beginning with the third line\,
a \<\< appears between columns of output that wasn't there before\.

Here's a testcase&#8203;:

\-\-\-\-\-

&nbsp;  \#\! /usr/bin/perl

&nbsp;  use strict;
&nbsp;  use warnings;
&nbsp;  use open '&#8203;:utf8';
&nbsp;  use open OUT => "&#8203;:encoding\(iso\-8859\-1\)";
&nbsp;  use Encode;

&nbsp;  open INFIL\, "test\.data";

&nbsp;  my \($author\,
&nbsp;      $title\);

&nbsp;  open \(TXTOUT\, ">test\.txt"\) or die "Cannot open output progs&#8203;: $\!\\n";

&nbsp;  while \(\<INFIL>\) \{
&nbsp;      chomp;
&nbsp;      \($author\, $title\) = split /\\|/;
&nbsp;      write TXTOUT;
&nbsp;  \}

&nbsp;  close INFIL;
&nbsp;  close TXTOUT;

&nbsp;  format TXTOUT =
&nbsp;  @&#8203;\<\<\<\<\<\<\<\<\<\<\<\<\<\<  @&#8203;\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<  @&#8203;\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<
&nbsp;  "1234567890123"\, $author\, $title
&nbsp;  \.

\-\-\-\-\-

And here's the data file that the above uses\.  It will probably get hopelessly mangled in delivery\.  There are TWO pipe\-delimited fields on every line\, although the first field is\, on occasion\, empty\.  There is a trailing pipe on each line\.  

\-\-\-\-\-

Grisham\, John|The chamber \[audiobook\] / John Grisham|
|Flic story \[DVD\] / Pathé Renn production ; StudioCanal Image ; Mondial Tefi ; a film by Jacques Deray ; screenplay by Jacques Deray and Alphonse Boudard ; directed by Jacques Deray|
Tempelman\-Kluit\, Anne\, 1941\-|Yukon / by Anne Tempelman\-Kluit ; consultants&#8203;: Desmond Morton\.\.\.\[et al\.\]|
Wood\, Tim|Motor racing / Tim Wood ; photographs\, Chris Fairclough|
Birdsall\, Jeanne|The Penderwicks : a summer tale of four sisters\, two rabbits\, and a very interesting boy / Jeanne Birdsall|
|Cosmopolitan \(New York\, N\.Y\. : 1952\)|
|The Performing arts and American society|
Dershowitz\, Alan M|Taking liberties : a decade of hard cases\, bad laws\, and bum raps / Alan M\. Dershowitz|
Hurd\, Edith Thacher\, 1910\-|The mother beaver\. Illustrated by Clement Hurd|
Williams\, Dyfri|Greek gold : jewelry of the classical world / Dyfri Williams and Jack Ogden|
Schultz\, Ken|Greatest fishing locales of North America / text and photography by Ken Schultz|
L'Engle\, Madeleine|A live coal in the sea / Madeleine L'Engle|

\-\-\-\-\-

If you need either of these files as an attachment\, let me know and I'll send\.

If I move the data into a \_\_DATA\_\_ block at the end of the source file\, I don't have this problem\.  Also\, at one point when testing\, I changed the spaces between columns to another character\, and they still changed to "\<\<"\.

I've also test this using the Sun\-supplied perl 5\.8\.4 on Solaris/SPARC and on a software vendor\-supplied 
gcc compiled perl 5\.8\.7 on Solaris/SPARC with the same results\.  I have not tested on perl 5\.10

I did find a workaround for this problem\.  If I use encode\(\) to explicitly convert text to Latin1 before calling
write\(\)\, I don't have this problem\.  Here's an example&#8203;:

$author = encode\("iso\-8859\-1"\, $author\);

Thanks

Michael Koehn

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.8:

Configured by SYSTEM at Tue Jul 31 19:34:29 2007.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=MSWin32, osvers=5.00, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -GF -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -DNO_HASH_SEED -DUSE_SITECUSTOMIZE -DPRIVLIB_LAST_IN_INC -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='12.00.8804', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"C:\Perl\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl58.lib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"C:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    Iin_load_module moved for compatibility with build 806
    Avoid signal flag SA_RESTART for older versions of HP-UX
    PerlEx support in CGI::Carp
    Less verbose ExtUtils::Install and Pod::Find
    Patch for CAN-2005-0448 from Debian with modifications
    Rearrange @INC so that 'site' is searched before 'perl'
    Partly reverted 24733 to preserve binary compatibility
    MAINT31223 plus additional changes
    31490 Problem bootstraping Win32CORE
    31324 Fix DynaLoader::dl_findfile() to locate .so files again
    31214 Win32::GetLastError fails when first called
    31211 Restore Windows NT support
    31188 Problem killing a pseudo-forked child on Win32
    29732 ANSIfy the PATH environment variable on Windows
    27527,29868 win32_async_check() can loop indefinitely
    26970 Make Passive mode the default for Net::FTP
    26379 Fix alarm() for Windows 2003
    24699 ICMP_UNREACHABLE handling in Net::Ping


@INC for perl v5.8.8:
    C:/Perl/site/lib
    C:/Perl/lib
    .


Environment for perl v5.8.8:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Perl\bin;C:\Perl\bin\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\Program Files\ATI Technologies\ATI Control Panel;C:\Program Files\Common Files\Adaptec Shared\System;c:\cygwin\bin;C:\Program Files\QuickTime\QTSystem\
    PERL_BADLANG (unset)
    SHELL (unset)





      ____________________________________________________________________________________
Be a better friend, newshound, and 
know-it-all with Yahoo! Mobile.  Try it now.  http://mobile.yahoo.com/;_ylt=Ahu06i62sR8HDtDypao8Wcj9tAcJ
```
</details>

**Comment by p5pRT** [comment]
### From @chorny

Hello\, porters\.

Same problem on 5\.10/Windows\. I'm attaching archive with program and text files\.
in\.txt \- with an accented character
in1\.txt \- without

2008/4/19\, via RT Michael Koehn \<perlbug\-followup@&#8203;perl\.org>&#8203;:
>  I'm reading Unicode data from a file\, and have set the output filehandle to be ISO\-8859\-1 \(Latin1\)\.
>  I'm using a format to format the output\.  Evertything is fine in the beginning\.  When I hit the first
>  line with an accented character\, the two spaces that I use as column dividers change into "\<\<" for the
>  rest of the program run\.

\-\- 
Alexandr Ciornii\, http&#8203;://chorny\.net

**Comment by p5pRT** [comment]
### From @chorny

[53054.zip](https://rt-archive.perl.org/perl5/Ticket/Attachment/389198/176498/53054.zip)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

This was fixed by commit f3f2f1a3479714dd7c55621da87bce4872a4ab17
Author&#8203;: Nicholas Clark \<nick@&#8203;ccl4\.org>
Date&#8203;:   Fri Jul 11 20&#8203;:04&#8203;:57 2008 \+0000

&nbsp;    Regression tests for\, and fix quite a lot of bugs in\, cases FF\_LINESNGL
&nbsp;    and FF\_LINEGLOB in pp\_formline\. \(While investigating RT \#55668\)
&nbsp;    Sometimes I wonder if this is actually fun\.
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'resolved'
